mkRemovedOptionModule: add replacement argument

Profpatsch opened the pull request:
When displaying a warning about a removed Option we should always
include reasoning why it was removed and how to get the same
functionality without it.

Introduces such a description argument and patches occurences (mostly
with an empty string).

Tested manually by trying the cases: empty string replacement message, nonempty string replacement message, with the `startGnuPGAgent` option.

cc @peti, I’ve copied your commit description for the gpg-agent and added a few changes (second commit).

Example:
![screenshot](https://cloud.githubusercontent.com/assets/3153638/14128415/1292a46c-f621-11e5-8caa-08e3811733ee.png)

Profpatsch committed:
modules/mkRemovedOptionModule: add replacement doc

When displaying a warning about a removed Option we should always
include reasoning why it was removed and how to get the same
functionality without it.

Introduces such a description argument and patches occurences (mostly
with an empty string).

startGnuPGAgent: further notes on replacement

peti commented on nixos/modules/rename.nix:
```diff
@@ -102,16 +102,47 @@ with lib;
+      agent" scheme we've implemented in NixOS for older versions.
+
+      To configure the gpg-agent for your X session, add the following code to
+      ~/.bashrc or some file that’s sourced when your shell is started:
```
`.bashrc` doesn't seem right for all of the commands mentioned below. There's certainly no point in running `gpg-connect-agent /bye` every time a shell is started.

Profpatsch replied:
Right, I didn’t read this correctly. The second block needs to be in `xinitrc`, only the first one should be per-shell.

peti commented on nixos/modules/rename.nix:
```diff
@@ -102,16 +102,47 @@ with lib;
+      If gpg’s automatic transformation of the private keys to the new format fails,
+      you will need to re-import your private keyring as well:
+
+          gpg --import ~/.gnupg/secring.gpg
```
Huh? Why would it be necessary to import the secret key ring into itself? That sounds like a no-op to me. If that is really necessary, then this is clearly a bug in GnuPG that should be reported upstream so that it can be fixed.

Profpatsch replied:
2.1 moved the location of the private keys, they are now managed by `gpg-agent`. See https://www.gnupg.org/faq/whats-new-in-2.1.html#nosecring

domenkozar commented:
Thanks!

domenkozar commented:
I'd appreciate if commits were split so that the gnupg text was separate from the functionality (easier for cherry-picking). But it's no biggie.

Profpatsch commented:
@domenkozar The first commit already copies part of it over from the commit message, so I thought it would be nice to have a consistent view.

Profpatsch commented on nixos/modules/rename.nix:
```diff
@@ -117,14 +117,15 @@ with lib;
 
-      If you want to use gpg-agent for SSH, too, also add the settings
+      If you want to use gpg-agent for SSH, too, add the following to your session
+      initialization (e.g. `displayManager.sessionCommands`)
```
@peti corrected it.

edolstra commented:
I'm not sure if it's a good idea to use the Nix trace mechanism for presenting large amounts of text, which more properly belong in the manual. I mean, I get annoyed by *single-line* trace messages, but getting a whole page of text every time I run nixos-rebuild would be very bad... Probably something like `Option foo has been removed; see https://nixos.org/nixos/manual/release-notes.html#sec-release-15.09 for more information` would be better.

edolstra commented on lib/modules.nix:
```diff
@@ -515,7 +520,14 @@ rec {
+
+            ${if replacementInstructions != ""
+              then replacementInstructions
+              else "No instructions were provided on how to replace its functionality."
```
This is superfluous; it's not necessary to tell the user we're not telling them something.

peti commented:
:+1: on adding these instructions to the release notes and linking there rather than putting the whole text inline.

Profpatsch commented:
> I mean, I get annoyed by single-line trace messages, but getting a whole page of text every time I run nixos-rebuild would be very bad...

Removing the option from the concerning config would fix that.
I don’t see why yet another online dependency is better than directly having the information directly available.

peti commented:
> I don’t see why yet another online dependency is better than directly having the information directly available.

Installation instructions for GnuPG are off-topic for the Nixpkgs git repository. I think it's a great service to our users to provide helpful error messages and hints for proper upgrade paths, etc., but quite a bit of the message you're suggesting is actually unrelated to Nixpks and  NixOS. The bit about re-importing the secret keyring, for instance, is totally unrelated to this option, even. I'm in favor of having this kind of helpful bits available, but the release notes do seem like a better place for this than the NixOS source tree.

peti commented:
Also, I'd like to echo @domenkozar's point from https://github.com/NixOS/nixpkgs/pull/14311#issuecomment-203388401. The current patch mixes two unrelated'ish changes. As it is, it's impossible to cherry-pick the NixOS extension to `release-16.03` without cherry-picking the deprecation of the agent option, too. But we didn't update GnuPG in the `release-16.03` branch; the agent option still exists there.

Profpatsch commented:
> Installation instructions for GnuPG are off-topic for the Nixpkgs git repository.

Ah, I understand.



> As it is, it's impossible to cherry-pick the NixOS extension to release-16.03 without cherry-picking the deprecation of the agent option, too.

I’m not sure how I should split them then, because I’d have to modify the gpg line in any case; would that lead to a merge conflict on cherry pick?

peti commented:
> I’m not sure how I should split them then, because I’d have to modify the gpg line in any case; would that lead to a merge conflict on cherry pick?

Oh, right. That's true. We cannot avoid a merge conflict because the file differs (significantly) in both branches already. :-( Well, I guess then there's no advantage to be had from splitting the first commit.

domenkozar commented:
@Profpatsch care to fix @edolstra comments so we can merge?

Profpatsch commented:
@domenkozar I think I addressed the comments.
I still don’t like the weak referencing of the manual. What if someone moves it into another section? There is no explicit connection. How can it be improved?

Profpatsch commented:
ping @domenkozar @peti

peti commented:
Please re-base this branch relative to the current head of `master`.

Profpatsch commented:
done cc @peti
